### PR TITLE
Added CKEDITOR.config.image2_maxSize

### LINKED ISSUE #2048: Image2 could have limiter for resizing image to specified maximum
## Type of report

Feature request

## Provide description of the new feature

Originally reported here: https://dev.ckeditor.com/ticket/12701

Image2 could have an extra setting or field in dialog for defining maximum resize size. If doing that inside the dialog, we could have a checkbox (Limit resizer to current image size) or 2 fields for entering max width/height manually. We could also have a configuration option with predefined values e.g. `image_size`, `image_size+20`, `200x300`.

### LINKED ISSUE #2672: image2: resizing to minimum size doesn't update dialogs values
## Type of report

Bug

## Provide detailed reproduction steps (if any)

1. Open test `http://localhost:1030/tests/plugins/image2/manual/resizer`
2. Scroll to last editor.
3. Resize image to minimum size.
4. Double click on image to open dialog.

### Expected result

- Dialogs fields match image properties.

### Actual result

- `width` and `height` are unset.

## Other details

Look at code:
https://github.com/ckeditor/ckeditor-dev/blob/7c937e875c35b9bd333fcc5de9c0910a2584ad34/plugins/image2/plugin.js#L1317-L1322

This is the part of `mousemove` handler. When mouse moves decreasing image size, attributes of img element are changed, but `widget.data` is updated in `mouseup` handler only if `updateData` is set to true. So in this scenario image size is changed, but the second one of dimensions is lower than allowed `updateData` is overwritten with `false` and thus data is not updated, and dialog shows previews value.

* Browser: …
* OS: …
* CKEditor version: …
* Installed CKEditor plugins: …

### LINKED ISSUE #2714: waitForImage should call callback if image is already loaded
## Type of report

Task

## Provide description of the task

Test tool `waitForImage` calls callback on `load` event, but when image is already loaded this event will never happen, and some browsers might load images earlier than others.

https://github.com/ckeditor/ckeditor-dev/blob/master/tests/plugins/uploadfile/_helpers/waitForImage.js

I propose to add if branch:

```
else if ( image.$.complete )  {
    callback()
}
```

## Other details

* Browser: …
* OS: …
* CKEditor version: …
* Installed CKEditor plugins: …

### PR DESCRIPTION
**Pull request by engineering-this**
## What is the purpose of this pull request?

New feature + Bugfix

## Does your PR contain necessary tests?

All patches which change the editor code must include tests. You can always read more
on [PR testing](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_contributing_code.html#tests),
[how to set the testing environment](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html) and
[how to create tests](https://ckeditor.com/docs/ckeditor4/latest/guide/dev_tests.html#creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [X] Unit tests
- [X] Manual tests

## What changes did you make?

- Added new config option `CKEDITOR.config.image2_maxSize` that limits maximum size that image can be resized with resizer.
- Fixed resizer `mousemove` listener to update widget data with last correct value which is same as one set on element.

Closes #2048 
Closes #2672
Closes #2714

### CODE REVIEW

File: tests/plugins/image2/resizehandler.js [hunk 1]
```diff
@@ -0,0 +1,145 @@
...
+	};
+
+	bender.test( bender.tools.createTestsForEditors( CKEDITOR.tools.objectKeys( bender.editors ), tests ) );
+
+	function testResize( options ) {
+		return function( editor, bot ) {
+			bot.setData( '<img src="%BASE_PATH%/_assets/logo.png">', function() {
+				var image = editor.editable().findOne( 'img' );
+
+				if ( CKEDITOR.env.chrome ) {
```
Comment (Comandeer) [inline tests/plugins/image2/resizehandler.js]: Is this branch necessary? For the sake of the simplicity, Chrome can also wait for image.

Comment (engineering-this) [reply tests/plugins/image2/resizehandler.js]: Actually at this point Chrome unlike other browsers has image loaded, and `waitForImage` triggers on `load` event which won't happen in Chrome (#2714).

File: plugins/image2/plugin.js [hunk 2]
```diff
@@ -1157,6 +1157,15 @@
 		resizer.on( 'mousedown', function( evt ) {
 			var image = widget.parts.image,
 
+				// Don't update attributes if less than 15.
+				// This is to prevent images to visually disappear.
+				min = {
```
[the remark this answers is not in the document]
Comment (engineering-this) [reply plugins/image2/plugin.js]: I don't think it's important to allow setting min size for resizer. Also it wasn't mentioned in original issue.

File: tests/plugins/image2/manual/dialogafterresize.md [hunk 3]
```diff
@@ -0,0 +1,15 @@
+@bender-tags: 4.12.0, feature, bug, 2672
```
Comment (Comandeer) [inline tests/plugins/image2/manual/dialogafterresize.md]: After splitting test into two separate ones, declaring them as `feature` & `bug` is no longer needed – one is for feature and one is for bug.

File: tests/plugins/image2/manual/dialogafterresize.md [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+@bender-ui: collapsed
+@bender-ckeditor-plugins: wysiwygarea, toolbar, image2, sourcearea
+
+1. Use resize handler to shrink image to minimum size.
+
+1. Open dialog for this image.
+
+## Expected:
+
+Dialog fields are populated with correct width/height.
```
Comment (Comandeer) [inline tests/plugins/image2/manual/dialogafterresize.md]: It's not clear, which values are correct and which are not.

File: tests/plugins/image2/resizehandler.js [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+				image2_maxSize: {
+					width: 'natural',
+					height: 'natural'
+				}
+			}
+		}
+	};
+
+	var tests = {
+		//2672
```
Comment (Comandeer) [inline tests/plugins/image2/resizehandler.js]: Our convention for such comments is `// (#<issue number>)`.

File: tests/plugins/uploadfile/uploadfile.js [hunk 6]
```diff
@@ -226,5 +226,21 @@ bender.test( {
 		loader.abort();
 
 		assert.areSame( 'abort', loader.status, 'Loader status' );
+	},
+
+	// #2714
+	'test waitForImage when image is loaded': function() {
```
Comment (Comandeer) [inline tests/plugins/uploadfile/uploadfile.js]: This test should not be here, as it's connected to the test helper, not to the plugin itself. In fact helpers usually don't have tests.

**Comment by engineering-this** [comment]
- Updated manual test.
- `waitForImage` unit test extracted to separate file.
- Made `waitForImage` callback call asynchronous when image is loaded early.

**Pull request merged**